src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,143 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private Object[] elements;
+    private int size = 0;
+    private int capacity = 10;
+
+    public ArrayList() {
+        this.elements = new Object[DEFAULT_CAPACITY];
+    }
+
     @Override
     public void add(T value) {
-
+        ensureCapacity(size + 1);
+        elements[size] = value;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkInsertIndex(index);
+        ensureCapacity(size + 1);
 
+        if (index < size) {
+            shiftRightFromIndex(index);
+        }
+
+        elements[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
+        ensureCapacity(size + list.size());
+
+        for (int i = 0; i < list.size(); i++) {
+            elements[size + i] = list.get(i);
+        }
 
+        size += list.size();
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T get(int index) {
-        return null;
+        checkAccessIndex(index);
+        return (T) elements[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        checkAccessIndex(index);
+        elements[index] = value;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T remove(int index) {
-        return null;
+        checkAccessIndex(index);
+
+        T removeElement = (T) elements[index];
+        shiftLeftFromIndex(index);
+        size--;
+        return removeElement;
     }
 
+    @SuppressWarnings("unchecked")
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == null) {
+                if (elements[i] == null) {
+                    shiftLeftFromIndex(i);
+                    size--;
+                    return null;
+                }
+            } else if (element.equals(elements[i])) {
+                T removed = (T) elements[i];
+                shiftLeftFromIndex(i);
+                size--;
+                return removed;
+            }
```

Comment:
move logic to get the element index to the private method and simplify it using `System.arraycopy`